example/no_tls_async_reconnect_client.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright Takatoshi Kondo 2020
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#include <iostream>
+
+#include <mqtt_client_cpp.hpp>
+
+// This example shows the client reconnecting to the broker
+//
+// Important: please note that messages are only republished to the broker if
+// the broker still has an active session for this client. Otherwise a clean
+// start is performed
+
+template<typename C>
+void reconnect_client(boost::asio::deadline_timer &timer, C &c)
+{
+    std::cout << "Start reconnect timer" << std::endl;
+
+    // Set an expiry time relative to now.
+    timer.expires_from_now(boost::posix_time::seconds(5));
+
+    timer.async_wait([&timer, &c](const boost::system::error_code& error) {
+        if (error != boost::asio::error::operation_aborted) {
+            std::cout << "Reconnect now !!" << std::endl;
+
+            // Connect
+            c->async_connect(
+                // [optional] checking underlying layer completion code
+                [&timer, &c]
+                (MQTT_NS::error_code ec){
+                    std::cout << "async_connect callback: " << ec.message() << std::endl;
+                    if (ec && ec != boost::asio::error::operation_aborted) {
+                        reconnect_client(timer, c);
+                    }
+                }
+            );
+        }
+    });
+}
+
+template<typename C>
+void publish_message(boost::asio::deadline_timer &timer, C &c, int packet_counter)
+{
+    // Publish a message every 5 seconds
+    timer.expires_from_now(boost::posix_time::seconds(5));
+
+    timer.async_wait([&timer, &c, packet_counter = packet_counter](const boost::system::error_code& error) {
+        if (error != boost::asio::error::operation_aborted) {
+            auto message = std::make_shared<std::string>("packet #" + std::to_string(packet_counter));
+            c->async_publish(
+                "mqtt_client_cpp/topic1",
+                *message,
```

Comment:
```suggestion
                MQTT_NS::allocate_buffer("packet #" + std::to_string(packet_counter))
```

*1
I think that the following overload is used.
https://github.com/redboltz/mqtt_cpp/blob/master/include/mqtt/endpoint.hpp#L2444-L2511
And lifetime is automatically managerd.
https://github.com/redboltz/mqtt_cpp/blob/master/include/mqtt/endpoint.hpp#L2504-L2508